### PR TITLE
[SLP] Extract isIdentityOrder to common routine [probably NFC]

### PR DESCRIPTION
**Pull request by preames**
This isn't quite just code motion as the four different versions we had of this routine differed in whether they ignored the "size" marker used to represent undef.  I doubt this matters in practice, but it is a functional change.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -1364,6 +1364,19 @@ class BoUpSLP {
   /// Perform LICM and CSE on the newly generated gather sequences.
   void optimizeGatherSequence();
 
+  /// Does this non-empty order represent an identity order?  Identity
+  /// should be represented as an empty order, so this is used to
+  /// decide if we can canonicalize a computed order.  Undef elements
+  /// (represented as size) are ignored.
+  bool IsIdentityOrder(ArrayRef<unsigned> Order) const {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:1371]: isIdentityOrder

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -1364,6 +1364,19 @@ class BoUpSLP {
...
+    assert(!Order.empty());
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:1372]: Add assertion message

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -1364,6 +1364,19 @@ class BoUpSLP {
...
+    const unsigned Sz = Order.size();
+    for (unsigned Idx : seq<unsigned>(0, Sz))
+      if (Idx != Order[Idx] && Order[Idx] != Sz)
+        return false;
+    return true;
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:1377]: `return all_of(enumerate(Order), [&](const auto &P) { return P.value() == P.index() || P.value() == Sz; });`

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 4]
```diff
@@ -1364,6 +1364,18 @@ class BoUpSLP {
   /// Perform LICM and CSE on the newly generated gather sequences.
   void optimizeGatherSequence();
 
+  /// Does this non-empty order represent an identity order?  Identity
+  /// should be represented as an empty order, so this is used to
+  /// decide if we can canonicalize a computed order.  Undef elements
+  /// (represented as size) are ignored.
+  bool isIdentityOrder(ArrayRef<unsigned> Order) const {
+    assert(!Order.empty() && "expected non-empty mask");
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:1372]:
```suggestion
    assert(!Order.empty() && "expected non-empty order");
```

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 5]
```diff
@@ -1364,6 +1364,18 @@ class BoUpSLP {
   /// Perform LICM and CSE on the newly generated gather sequences.
   void optimizeGatherSequence();
 
+  /// Does this non-empty order represent an identity order?  Identity
+  /// should be represented as an empty order, so this is used to
+  /// decide if we can canonicalize a computed order.  Undef elements
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:1369]: I don't quite understand the description. Why it is empty order?

Comment (preames) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:1369]: This is just describing the behavior of the existing code.  We use an empty order to represent no reordering - i.e. the identity order.  We use this routine to check if a computed order is identity, and can thus be replaced with an empty one.  If you have a preferred wording here, please suggest it.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:1369]: Ah, I see

**Review by alexey-bataev — APPROVED** [review]

**Pull request merged**